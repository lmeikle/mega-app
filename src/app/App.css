.app {
  max-width: 400px;
  margin-left: auto;
  margin-right: auto;
  text-align: center;
  font-family: 'Titillium Web', sans-serif;
  color: #002742;
  padding: 10px;
}

a {
  text-decoration: none;
}
a {
  color: #1e70bf;
}
a:hover {
  text-decoration: none;
}
a:hover {
  color: #ed3773;
}
