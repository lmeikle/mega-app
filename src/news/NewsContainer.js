import React, { Component } from 'react';
import { fetchTopHeadlines } from './NewsAPI';
import NewsHeadlineComponent from './NewsHeadlineComponent';

/**
 * Renders a list of top news headlines
 */
class NewsContainer extends Component {
  constructor(props) {
    super(props);

    this.state = {
      totalResults: 0,
      headlines: [],
      error: null
    };
  }

  static get PAGE_SIZE() {
    return 5;
  }

  componentDidMount() {
    this.showMore();
  }

  showMore = () => {
    let { totalResults, headlines } = this.state;
    let page = headlines.length / totalResults * totalResults / NewsContainer.PAGE_SIZE;
    fetchTopHeadlines(page + 1, NewsContainer.PAGE_SIZE)
      .then(result =>
        this.setState({
          headlines: [...this.state.headlines, ...result.headlines],
          totalResults: result.totalResults
        })
      )
      .catch(error => this.setState({ error }));
  };

  render() {
    const { headlines, totalResults, error } = this.state;

    if (error) {
      return <div className="errorMessage">Failed to get top headlines due to: {error.toString()}</div>;
    }

    return (
      <div className="news-container">
        <div className="news-title">News</div>
        <div className="news-headlines-container">
          {headlines.map(headline => <NewsHeadlineComponent key={headline.title} {...headline} />)}
        </div>
        {headlines.length < totalResults && (
          <div className="news-show-more-button" onClick={this.showMore}>
            Show More
          </div>
        )}
        <div className="news-powered-by">
          <a href="https://newsapi.org" target="_blank" rel="noopener noreferrer">
            Powered by News API
          </a>
        </div>
      </div>
    );
  }
}

export default NewsContainer;
