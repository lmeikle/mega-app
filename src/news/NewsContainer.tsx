import * as React from 'react';
import { LoadingComponent } from '@lmeikle/my-mono-repo-to-single-package';
import { fetchTopHeadlines } from './NewsAPI';
import NewsHeadlineComponent from './NewsHeadlineComponent';

type IProps = {};

type IState = {
  totalResults: number;
  headlines: Array<IHeadLineProps>;
  error?: string;
};

export type IHeadLineProps = {
  title: string;
  date: string;
  source: string;
  url: string;
};

/**
 * Renders a list of top news headlines
 */
class NewsContainer extends React.Component<IProps, IState> {
  state: IState = {
    totalResults: 0,
    headlines: []
  };

  static get PAGE_SIZE() {
    return 5;
  }

  componentDidMount() {
    this.showMore();
  }

  showMore = async () => {
    const { totalResults, headlines } = this.state;
    let page = ((headlines.length / totalResults) * totalResults) / NewsContainer.PAGE_SIZE;

    try {
      const result = await fetchTopHeadlines(page + 1, NewsContainer.PAGE_SIZE);
      this.setState({
        headlines: [...this.state.headlines, ...result.headlines],
        totalResults: result.totalResults
      });
    } catch (error) {
      this.setState({ error });
    }
  };

  render() {
    const { headlines, totalResults, error } = this.state;

    if (error) {
      return <div className="errorMessage">Failed to get top headlines due to: {error.toString()}</div>;
    }

    if (totalResults === 0) {
      return <LoadingComponent />;
    }

    return (
      <div className="news-container">
        <div className="news-title">News</div>
        <div className="news-headlines-container">
          {headlines.map(headline => (
            <NewsHeadlineComponent key={headline.title} {...headline} />
          ))}
        </div>
        {headlines.length < totalResults && (
          <div className="news-show-more-button" onClick={this.showMore}>
            Show More
          </div>
        )}
        <div className="news-powered-by">
          <a href="https://newsapi.org" target="_blank" rel="noopener noreferrer">
            Powered by News API
          </a>
        </div>
      </div>
    );
  }
}

export default NewsContainer;
