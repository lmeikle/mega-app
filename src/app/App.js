import React, { Component } from "react"
import { BrowserRouter as Router, Route, Switch } from "react-router-dom"
import Banking from "../banking/page/Banking"
import Home from "../home/Home"
import "./App.css"

class App extends Component {
  render() {
    return (
      <div className="app">
        <Router>
          <Switch>
            <Route path="/banking" component={Banking} />
            <Route path="/" component={Home} />
          </Switch>
        </Router>
      </div>
    )
  }
}

export default App
