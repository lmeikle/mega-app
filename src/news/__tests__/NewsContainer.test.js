import React from 'react';
import { mount, shallow } from 'enzyme';
import NewsContainer from '../NewsContainer';
import App from '../../app/App';

const initialState = {
  totalResults: 0,
  headlines: [],
  error: null
};

const readyState = {
  totalResults: 2,
  headlines: [
    {
      title: 'Putin ready to meet Trump over arms race concerns',
      date: '10/06/2018',
      source: 'CNN',
      url: 'https://www.cnn.com/2018/06/10/politics/putin-trump-us-russia-arms-intl/index.html'
    },
    {
      title: 'Leave.EU downplayed Russian links',
      date: '10/06/2018',
      source: 'Bbc.com',
      url: 'https://www.bbc.com/news/uk-politics-44428115'
    }
  ],
  error: null
};

const errorState = {
  error: 'There was an error'
};

const newsContainer = shallow(<NewsContainer />);

describe('NewsContainer', () => {
  test('renders correctly', () => {
    expect(newsContainer).toMatchSnapshot();
  });

  test('initializes the `state` correctly', () => {
    expect(newsContainer.state()).toEqual(initialState);
  });

  test('renders a list of NewsHeadlineComponents', () => {
    newsContainer.setState(readyState);
    expect(newsContainer.find('NewsHeadlineComponent').length).toBe(2);
  });

  test('displays an error message', () => {
    newsContainer.setState(errorState);
    expect(newsContainer.find('.errorMessage').exists()).toBe(true);
  });
});
