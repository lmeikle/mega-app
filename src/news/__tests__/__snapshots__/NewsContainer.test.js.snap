// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewsContainer renders correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <NewsContainer />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="news-title"
>
          News
</div>,
        <div
          className="news-headlines-container"
>
          
</div>,
        false,
        <div
          className="news-powered-by"
>
          <a
                    href="https://newsapi.org"
                    rel="noopener noreferrer"
                    target="_blank"
          >
                    Powered by News API
          </a>
</div>,
      ],
      "className": "news-container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "News",
          "className": "news-title",
        },
        "ref": null,
        "rendered": "News",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [],
          "className": "news-headlines-container",
        },
        "ref": null,
        "rendered": Array [],
        "type": "div",
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": <a
            href="https://newsapi.org"
            rel="noopener noreferrer"
            target="_blank"
>
            Powered by News API
</a>,
          "className": "news-powered-by",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Powered by News API",
            "href": "https://newsapi.org",
            "rel": "noopener noreferrer",
            "target": "_blank",
          },
          "ref": null,
          "rendered": "Powered by News API",
          "type": "a",
        },
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="news-title"
>
            News
</div>,
          <div
            className="news-headlines-container"
>
            
</div>,
          false,
          <div
            className="news-powered-by"
>
            <a
                        href="https://newsapi.org"
                        rel="noopener noreferrer"
                        target="_blank"
            >
                        Powered by News API
            </a>
</div>,
        ],
        "className": "news-container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "News",
            "className": "news-title",
          },
          "ref": null,
          "rendered": "News",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [],
            "className": "news-headlines-container",
          },
          "ref": null,
          "rendered": Array [],
          "type": "div",
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <a
              href="https://newsapi.org"
              rel="noopener noreferrer"
              target="_blank"
>
              Powered by News API
</a>,
            "className": "news-powered-by",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Powered by News API",
              "href": "https://newsapi.org",
              "rel": "noopener noreferrer",
              "target": "_blank",
            },
            "ref": null,
            "rendered": "Powered by News API",
            "type": "a",
          },
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
    "disableLifecycleMethods": true,
  },
}
`;
